clap_derive/src/derives/args.rs

```diff
@@ -51,10 +52,188 @@ pub(crate) fn derive_args(input: &DeriveInput) -> Result<TokenStream, syn::Error
+
+        let conflicts = variants
+            .iter()
+            .filter_map(|(_, v)| {
+                if v.ident == variant.ident {
+                    None
+                } else {
+                    Some(Name::Derived(v.ident.clone()))
+                }
+            })
+            .collect::<Vec<_>>();
+
+        let fields = collect_args_fields(item, fields)?;
+
+        let augmentation = gen_augment(&fields, &app_var, item, &conflicts, false)?;
+        let augmentation = quote! {
+            let #app_var = #augmentation;
+        };
+        let augmentation_update = gen_augment(&fields, &app_var, item, &conflicts, true)?;
+        let augmentation_update = quote! {
+            let #app_var = #augmentation_update;
+        };
+
+        augmentations.extend(augmentation);
+        augmentations_update.extend(augmentation_update);
+
+        let variant_name = &variant.ident;
+        let genned_constructor = gen_constructor(&fields)?;
+        let constructor = quote! {
+            if __clap_arg_matches.contains_id(#group_id) {
+                let v = #item_name::#variant_name #genned_constructor;
+                return ::std::result::Result::Ok(v)
+            }
+        };
+
+        constructors.extend(constructor);
+
+        let genned_updater = gen_updater(&fields, false)?;
+
+        let field_names = fields
+            .iter()
+            .map(|(field, _)| field.ident.as_ref().unwrap());
+        let updater = quote! {
+
+            if __clap_arg_matches.contains_id(#group_id) {
+                let #item_name::#variant_name { #( #field_names ),* } = self else {
+                    unreachable!();
+                };
+                #genned_updater;
+            }
+        };
+
+        updaters.extend(updater);
+    }
+
+    let raw_deprecated = raw_deprecated();
+
+    Ok(quote! {
+        #[allow(
+            dead_code,
+            unreachable_code,
+            unused_variables,
+            unused_braces,
+            unused_qualifications,
+        )]
+        #[allow(
+            clippy::style,
+            clippy::complexity,
+            clippy::pedantic,
+            clippy::restriction,
+            clippy::perf,
+            clippy::deprecated,
+            clippy::nursery,
+            clippy::cargo,
+            clippy::suspicious_else_formatting,
+            clippy::almost_swapped,
+            clippy::redundant_locals,
+        )]
+        #[automatically_derived]
+        impl #impl_generics clap::FromArgMatches for #item_name #ty_generics #where_clause {
+            fn from_arg_matches(__clap_arg_matches: &clap::ArgMatches) -> ::std::result::Result<Self, clap::Error> {
+                Self::from_arg_matches_mut(&mut __clap_arg_matches.clone())
+            }
+
+            fn from_arg_matches_mut(__clap_arg_matches: &mut clap::ArgMatches) -> ::std::result::Result<Self, clap::Error> {
+                #raw_deprecated
+                #constructors
+                unreachable!()
```

Comment:
we actually _do_ get here if neither variant was explcitly constructed.
In a way that makes sense, if all variants have defaultable arguments, either is a valid candidate. However I cant decide if that's a conflict or resolvable by defining a default varaint for such situations

Reply:
If the user has an `Option<Enum>`, we'll check if the group is present and do `None` if it isn't, so this is only a problem with `Enum`.  I lean towards hand constructing a clap error like we do if you have `#[arg(required = False)] String`.  If we make unit variants a fallback (or add a fallback attribute). then users can avoid the error here.